server/src/main/java/org/opensearch/index/mapper/BytesRefsCollectionBuilder.java

```diff
@@ -0,0 +1,178 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.index.mapper;
+
+import org.apache.lucene.search.TermInSetQuery;
+import org.apache.lucene.util.BytesRef;
+
+import java.util.AbstractSet;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Comparator;
+import java.util.Iterator;
+import java.util.List;
+import java.util.SortedSet;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+/**
+ * Purposed for passing terms into {@link TermInSetQuery}.
+ * If the given terms are sorted already, it wrap it with a SortedSet stub.
+ * Otherwise, it passes terms as list.
+ */
+public class BytesRefsCollectionBuilder implements Consumer<BytesRef>, Supplier<Collection<BytesRef>> {
+
+    /**
+     * Strategy for building BytesRef collection.
+     * */
+    protected interface CollectorStrategy extends Function<BytesRef, CollectorStrategy>, Supplier<Collection<BytesRef>> {}
+
+    protected final List<BytesRef> terms = new ArrayList<>();
+    protected CollectorStrategy delegate = createStartStrategy();
+
+    @Override
+    public void accept(BytesRef bytesRef) {
+        delegate = delegate.apply(bytesRef);
+    }
+
+    @Override
+    public Collection<BytesRef> get() {
+        Collection<BytesRef> result = delegate.get();
+        delegate = createFrozenStrategy(result);
+        return result;
+    }
+
+    protected CollectorStrategy createStartStrategy() {
+        return new CollectorStrategy() {
+            @Override
+            public CollectorStrategy apply(BytesRef firstBytes) {
+                terms.add(firstBytes); // firstly, just store
+                return createSortedStrategy(firstBytes);
+            }
+
+            @Override
+            public Collection<BytesRef> get() {
+                return terms; // empty list
+            }
+        };
+    }
```

Comment:
We can probably remove this. we can initialize the delegate while accepting `firstBytes`?

Reply:
I added the method to allow to override it for descentant.